### ISSUE TITLE
Pinball Fantasies (SVGA): wrong colors in table selection + cropped tables

### ISSUE THREAD

**Issue by carlosefr** [schellingb/dosbox-pure#553]
Opening an issue specifically for this, as it seems like something that won't have a quick fix, and to avoid the discussion being split between issue #550 and PR #552.

**Summary (from what was said before):**

Pinball Fantasies has historically required `vgaonly` in DOSBox. Right now, under SVGA the table selection screen shows the wrong colors (the 16-color palette change for the bottom half isn't applied), and the tables themselves appear cropped at the bottom.

Even though the table selection screen is 640x480 (`vga.mode=2`), which doesn't require double-scanning in VGA, the game still sets double-scanning explicitly in the [maximum scan line register](http://www.osdever.net/FreeVGA/vga/crtcreg.htm#09) (`SD=1`).

In `vgaonly` the SD bit in that register is ignored for `vga.mode` 2 and 3 — [here](https://github.com/schellingb/dosbox-pure/blob/39dd77adff71cd2ae17d80a0e48fad8583ade8e4/src/hardware/vga_draw.cpp#L1154-L1162) — which is then expected to be compensated later on — [here](https://github.com/schellingb/dosbox-pure/blob/7f106c9abb2ab9f094c346c29e32f4f32d5e5b98/src/hardware/vga_draw.cpp#L1542-L1550).

In SVGA, the SD bit is honored — [here](https://github.com/schellingb/dosbox-pure/blob/39dd77adff71cd2ae17d80a0e48fad8583ade8e4/src/hardware/vga_draw.cpp#L1163) — and later causes the reported horizontal resolution to be halved — [here](https://github.com/schellingb/dosbox-pure/blob/39dd77adff71cd2ae17d80a0e48fad8583ade8e4/src/hardware/vga_draw.cpp#L1533-L1539).

Looking at the code in DOSBox-X and DOSBox-Staging doesn't help much, has `vga_draw.cpp` has been changed beyond recognition in those forks.

**Comment by carlosefr** [comment]
Lemmings also uses the same trick for the bottom control area. It also requires `vgaonly` for the level briefing screen, but the reason for that seems to be different (it doesn't require #552 to work).

**Comment by alexb3d** [comment]
There are many games that only work with VGA.

SVGA is just the standard, it is compatible with VGA but does not support all its features, it is not a video card compatible with both standards, VGA/SVGA.

**Comment by carlosefr** [comment]
@alexb3d SVGA cards were backwards compatible with VGA but, even if they weren't, that would not be relevant for this issue. The reason why DOSBox requires `vgaonly` for some games is due to limitations in DOSBox's code.

I've spent a few more hours on this after opening this issue and got nowhere. My conclusion is that the way `vga_draw.cpp` is written doesn't allow fixing this issue without a massive rewrite, there are just too many workarounds on top of other workarounds.

In summary, `vgaonly` is required for Pinball Fantasies to avoid breaking the default "SVGA" code paths in other ways, not because Pinball Fantasies depends on behavior that only original VGA cards implemented. In fact, `vgaonly` does not mean original VGA behavior at all, despite the name.

**Comment by carlosefr** [comment]
@schellingb Continuing here, since #552 has been merged...

> Maybe we can look into what DOSBox Staging did, it looks like Staging has gotten rid of vgaonly as a machine altogether, having only the SVGA cards above EGA.

They rewrote the whole thing, and seem to rely more closely to what's set in the VGA registers to derive the correct behavior. That's the correct approach.

Roughly, what Pinball Fantasies is doing here is setting things up in a way that triggers an interrupt midway through the frame as if it were starting a new frame (and that's why the resolution is reported as 640x320 instead of the correct 640x480). At that point it switches the palette, but the draw continues from the same position without doing a VSYNC.

[Line Counter Register](http://www.osdever.net/FreeVGA/vga/crtcreg.htm#18):

_"When the scan line counter reaches the value in the Line Compare field, the current scan line address is reset to 0 and the Preset Row Scan is presumed to be 0."_

I see places in the code where VGA triggers interrupts, and I see some references to the line counter. But can't, for the life of me, figure out where the two are glued together regarding this specific behavior. Nor where it is triggering a different behavior based on `vgaonly`. I've messed about with this, getting this mode to behave similarly to `vgaonly` on the double scanning side, but it still doesn't trigger the palette switching.

**Comment by johnnovak** [comment]
> They rewrote the whole thing, and seem to rely more closely to what's set in the VGA registers to derive the correct behavior. That's the correct approach.

As the guy who rewrote the whole thing in Staging, I agree :sunglasses: Only took me 100+ hours and lots of regression fixing. But it's the only way, yeah, and `vgaonly` is terrible hack in 2025 (it was necessary in the early 2000s for performance reason).
